packages/react-tooltip/src/components/Tooltip/Tooltip.types.ts

```diff
@@ -101,6 +136,32 @@ export type TooltipTriggerProps = Pick<
   'onPointerEnter' | 'onPointerLeave' | 'onFocus' | 'onBlur' | 'aria-describedby' | 'aria-labelledby' | 'aria-label'
 >;
 
+/**
+ * Data for the Tooltip's onBeforeShow event.
+ */
+export interface OnBeforeShowTooltipData {
+  /**
+   * The element that the tooltip will point to. By default, this is the element that triggered the tooltip
+   * (event.currentTarget). It can be changed by the onBeforeShow handler to have the tooltip point to another element.
+   */
+  target: HTMLElement;
+
+  /**
+   * Set preventShow to true to cancel showing the tooltip.
+   */
+  preventShow?: boolean;
+}
+
+/**
+ * Data for the Tooltip's onShow event.
+ */
+export interface OnShowTooltipData {
+  /**
+   * The element that the tooltip is pointing to.
+   */
+  readonly target: HTMLElement;
```

Comment:
`target` can also be a prop, this is done with controlled `open` boolean value for `Menu` and `Popover` already, there are stories too that show that pattern in action.

Reply:
This is the main concern for me - the API should be aligned across the components. It would be great if we could stick to either lifecycle methods or controlled components - or have a predictable guidelines on when which one is used.

Reply:
I'll change it to be consistent with menu/popover. I originally avoided a direct `target` prop since it requires a double-render to work because the target ref gets resolved _after_ the initial render. However, I like the consistency with menu/popover. And I expect the use of the `target` prop to be relatively uncommon; in the common case, the Tooltip's trigger will be used as the target, and that doesn't require a double-render to work.